src/Cryptol/Eval/Reference.lhs

```diff
@@ -540,21 +501,21 @@ To evaluate a declaration in a non-recursive context, we need only
 evaluate the expression on the right-hand side or look up the
 appropriate primitive.
 
-> evalDecl :: Env -> Decl -> (Name, Value)
+> evalDecl :: Env -> Decl -> (Name, E Value)
 > evalDecl env d =
 >   case dDefinition d of
->     DPrim   -> (dName d, evalPrim (dName d))
+>     DPrim   -> (dName d, pure (evalPrim (dName d)))
 >     DExpr e -> (dName d, evalExpr env e)
 
 To evaluate a declaration in a recursive context, we must perform a
 type-directed copy to build the spine of the value. This ensures that
 the definedness invariant for type `Value` will be maintained.
 
-> evalDeclRecursive :: Env -> Decl -> (Name, Value)
+> evalDeclRecursive :: Env -> Decl -> (Name, E Value)
 > evalDeclRecursive env d =
 >   case dDefinition d of
->     DPrim   -> (dName d, evalPrim (dName d))
->     DExpr e -> (dName d, copyBySchema env (dSignature d) (evalExpr env e))
+>     DPrim   -> (dName d, pure (evalPrim (dName d)))
+>     DExpr e -> (dName d, evalExpr env e)
```

Comment:
It looks like this change makes `evalDeclRecursive` totally identical to `evalDecl`, is that right? So we should probably just remove it (as well as the text above about the type-directed copy).

Reply:
Yeah, I'm still figuring out the details, and I'm not sure if there's any reason for these to be different anymore.  It seems like probably not.

Reply:
The only reason that `evalDeclRecursive` needed the type-directed copy was that our semantics required bottoms to occur only in the bit/integer leaves of the values. Without it, least fixed points would have introduced bottoms where they weren't supposed to be.

Now that evaluation returns something of type `E Value`, this is no longer a problem, as we explicitly allow bottoms at that type in the new semantics.

Actually, `evalDecl` and `evalDeclRecursive` never really needed to be different, as it would have been correct for `evalDecl` to also include the same type-directed copy. I just skipped it in the non-recursive case for efficiency reasons.